# Exploring Hacker News Post

today, i will exploring hacker news dataset,

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

hacker news dataset is summary of the hacker newst post, which is contain many information, such as number point of upvote, title, url, number of comments, author and many more.

This dataset will be analyzed to know more about pattern in dataset and also answer many question, such as Do posts created at a certain time receive more comments on average?



## importing dataset

In [1]:
#function to import dataset
from csv import reader

#open and read data using open and read function
def read_data(dataset, header = True):
    opened_data = open(dataset)
    reader_data = reader(opened_data)
    
    #make data become list
    new_data = list(reader_data)
    
    #print header in distinct list
    if header :
        return new_data[0], new_data[1:]
    
    return new_data

data_hacker_header, data_hacker = read_data('hacker_news.csv')

data_hacker[0:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

In [2]:
data_hacker_header

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

## Extracting Ask HN and Show HN Posts

 for this analyst we focus on post with title Ask Hn adn Show Hn,
 
 1. Ask Hn  are the posts to ask the Hacker News community a specific question, for instance "Ask HN: How to improve my personal website?"
 2. Show Hn are the posts to show the Hacker News community a project, product, or just generally something interesting

In [3]:
#extract row with ask hn and show hn
ask_posts = []
show_posts = []
other_posts = []

#loop the data
for row in data_hacker:
    title = row[1]
    title_lower = title.lower() #change title become lower, to make easy for analyst
    if title_lower.startswith('ask hn'):
        ask_posts.append(row)
    elif title_lower.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


from function above, i find that 1744 post begin with ask_posts and 1162 post begin with show_posts,



## Finding the Amount of Ask Posts and Comments by Hour

in this section, we'll determine if ask posts created at a certain time are more likely to attract comments
there are two steps to get the analysis:
1. find total number of comment per hour
2. find average amount of comment per hour

but, before we do analysis let's find total number comments between ask_posts and show_posts

In [17]:
#find total number of comment in ask_posts
total_ask_comments = 0

#loop the data
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments

#calculate average
avg_ask_comments = total_ask_comments / len(ask_posts)

print('ask_comments')
print('total_ask_comments: ', total_ask_comments)
print('avg_ask_comments: ' , avg_ask_comments)
print('\n')


#find total number of comment in show_posts
total_show_comments = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments
    
avg_show_comments = total_show_comments /len(show_posts)

print('shows_comments')
print('total_show_comments: ', total_show_comments)
print('avg_show_comments: ', avg_show_comments)
print('\n')
    

ask_comments
total_ask_comments:  24483
avg_ask_comments:  14.038417431192661


shows_comments
total_show_comments:  11988
avg_show_comments:  10.31669535283993




from my findings, people are likely more comments on ask posts rather than shows comments, let's see more deeply in ask_posts using 2 point which are created in the beginning of section

In [5]:
#amount of ask posts and comment per hours
import datetime as dt

#make empty list
result_list = []

#loop ask_posts
for row in ask_posts:
    created_at = row[6]
    n_comments = int(row[4])
    result_list.append([created_at, n_comments])
    
#make dictionary for count comment per hours
counts_by_hour = {}
comments_by_hour = {}
date_format = '%m/%d/%Y %H:%M'

#loop the data
for row in result_list:
    date = row[0]
    number_comments = int(row[1])
    date_strip = dt.datetime.strptime(date, date_format) #change the date using datetime function
    hour = date_strip.strftime('%H')
    
    #loop the data
    if hour not in counts_by_hour :
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = number_comments
    elif hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += number_comments

        


calculating the average comments per hour

In [8]:
#calculating average 

avg_by_hour = []

for data in counts_by_hour:
    avg_by_hour.append([data, comments_by_hour[data]/counts_by_hour[data]])

to make easy analysis, sorted the data

In [7]:
#swap the data
swap_avg_by_hour = []
for rows in avg_by_hour:
    swap_avg_by_hour.append([rows[1], rows[0]])
 
#sorted the data
sorted_swap = sorted(swap_avg_by_hour, reverse = True)

print('top 5 hours for ask post comments')

for avg, hour in sorted_swap[0:5]:
    date_strip = dt.datetime.strptime(hour, '%H')
    format_hour = date_strip.strftime('%H:%M:')
    average = avg
    print('{0} {1:.2f} average comments per post'.format(format_hour, average))

top 5 hours for ask post comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


from this analyst, we know that highest average comment in ask post it's happen in the afternoon, between 15:00 and 16:00 and in the evening, between 20:00 - 21:00 and 02:00.

if we compare with the office hours, most of people comment in ask section ehen people doesn't have working hours such as when in the road or in the cofee

In [11]:
data_hacker_header

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

## Finding the average point of ask_post and show_post

 i will try to finding the average point of both post and compare it

In [20]:
#average ask_post
total_ask_points = 0

#loop the data
for row in ask_posts:
    num_points = float(row[3])
    total_ask_points += num_points

#calculate average
avg_ask_points = total_ask_points / len(ask_posts)

print('ask_points')
print('total_ask_points : ', total_ask_points)
print('avg_ask_points : ', avg_ask_points)
print('\n')


#find total number of points in show_posts
total_show_points = 0
for row in show_posts:
    num_points = float(row[3])
    total_show_points += num_points
    
avg_show_points = total_show_points /len(show_posts)

print('shows_points')
print('total_show_points: ', total_show_points)
print('avg_show_points: ', avg_show_points)
print('\n')
    

ask_points
total_ask_points :  26268.0
avg_ask_points :  15.061926605504587


shows_points
total_show_points:  32019.0
avg_show_points:  27.555077452667813




from function above, i find that show_post have higher points than ask_points, 

for the next analysis, i will find average and total points between 2 posts

## Finding the Amount of Ask_posts points by Hour

same as comment section, i will try finding the amount and average of both posts

In [24]:
#make function to find top 5 amount and average of points

def top_5(dataset, name_dataset):
    import datetime as dt

    #make empty list
    result_list = []

    #loop dataset
    for row in dataset:
        created_at = row[6]
        n_points = float(row[3])
        result_list.append([created_at, n_points])
    
    #make dictionary for count comment per hours
    counts_by_hour = {}
    points_by_hour = {}
    date_format = '%m/%d/%Y %H:%M'

    #loop the data
    for row in result_list:
        date = row[0]
        number_points = float(row[1])
        date_strip = dt.datetime.strptime(date, date_format) #change the date using datetime function
        hour = date_strip.strftime('%H')
    
        #loop the data
        if hour not in counts_by_hour :
            counts_by_hour[hour] = 1
            points_by_hour[hour] = number_points
        elif hour in counts_by_hour:
            counts_by_hour[hour] += 1
            points_by_hour[hour] += number_points
            
    #calculating average 
    avg_by_hour = []

    for data in counts_by_hour:
        avg_by_hour.append([data, points_by_hour[data]/counts_by_hour[data]])
        
    #swap the data
    swap_avg_by_hour = []
    for rows in avg_by_hour:
        swap_avg_by_hour.append([rows[1], rows[0]])
 
    #sorted the data
    sorted_swap = sorted(swap_avg_by_hour, reverse = True)

    print('top 5 hours for {0} dataset'.format(name_dataset))

    for avg, hour in sorted_swap[0:5]:
        date_strip = dt.datetime.strptime(hour, '%H')
        format_hour = date_strip.strftime('%H:%M:')
        average = avg
        print('{0} {1:.2f} average points per post'.format(format_hour, average))
            
    

In [25]:
top_5(show_posts, 'show_posts')

top 5 hours for show_posts dataset
23:00: 42.39 average points per post
12:00: 41.69 average points per post
22:00: 40.35 average points per post
00:00: 37.84 average points per post
18:00: 36.31 average points per post


In [27]:
top_5(ask_posts, 'ask_posts')

top 5 hours for ask_posts dataset
15:00: 29.99 average points per post
13:00: 24.26 average points per post
16:00: 23.35 average points per post
17:00: 19.41 average points per post
10:00: 18.68 average points per post


we find that in show_posts, the high average points was happend in night and afternoon and it different with ask_posts.

in ask_post, most points came from afternoon time.

for the conclusion, all analysis give insight that many of people read post in hacker news in the time when they're not work or out of working hour